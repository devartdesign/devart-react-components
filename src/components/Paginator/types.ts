export enum LinkType {
  BUTTON,
  LINK
}

export interface IPaginatorQuery {
  page: number;
  limit: number;
}
