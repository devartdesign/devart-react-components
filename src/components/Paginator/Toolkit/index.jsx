export { default } from './Toolkit.container';
