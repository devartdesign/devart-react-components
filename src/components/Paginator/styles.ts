import { StyleSheet } from 'aphrodite/no-important';

export default StyleSheet.create({
  paginationNav: {
    display: 'flex',
    flexDirection: 'row',
    flexBasis: '100%',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  pageLimiter: {
    display: 'flex',
    width: '65px',
    height: '40px',
    backgroundColor: '#fff',
    borderRadius: '6px',
    boxShadow: '0 0 5px rgba(0, 0, 0, 0.1)',
  },
  navPagination: {
    display: 'flex',
    flexBasis: '100%;',
  },
  pagination: {
    backgroundColor: '#FFF',
    padding: '0 10px',
    margin: '10px 0',
    display: 'flex',
    flexGrow: 1,
    listStyleType: 'none',
  },
  pageItem: {
    display: 'flex',
    justifyContent: 'center',
    alignContent: 'center',
    width: 40,
    height: 25,
  },
  pageItemButton: {
    cursor: 'pointer',
    border: 'none',
    backgroundColor: '#FFF',
    width: '100%',
    height: '100%',
    padding: '0',
    ':hover': {
      backgroundColor: '#546e7a',
      color: '#FFF',
    }
  },
  pageItemActive: {
    backgroundColor: '#29434e',
    color: '#FFF',
  }
});
