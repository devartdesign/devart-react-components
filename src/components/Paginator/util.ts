export interface IPaginationRange {
  from: number;
  to: number;
  step?: number;
}

export const createPaginationRange = ({ from, to, step = 1 }: IPaginationRange) => {
  let i = from;
  const rangeNumber = [];
  while (i <= to) {
    rangeNumber.push(i);
    i += step;
  }
  return rangeNumber;
};
