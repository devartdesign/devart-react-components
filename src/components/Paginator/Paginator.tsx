import React, { memo, useMemo, useCallback, useState, useEffect } from 'react';
import { NavLink } from 'react-router-dom';
import { css } from 'aphrodite/no-important';

import { LinkType, IPaginatorQuery } from './types';
import { createPaginationRange } from './util';

import styles from './styles';

const LEFT_PAGE = 'LEFT';
const RIGHT_PAGE = 'RIGHT';

const PaginationButtonLink = params => (
  <button type="button" className={css(styles.pageItemButton)} aria-label="Next" onClick={params.onClick}>
    {!params.hideTick && (<span aria-hidden="true">&laquo;</span>)}
    <span>{params.text}</span>
  </button>
);

const PaginationNavLink = params => (
  <NavLink to={`/page/${params.page}`}>
    <span>{params.text}</span>
  </NavLink>
);

const typeMap = {
  [LinkType.LINK]: PaginationNavLink,
  [LinkType.BUTTON]: PaginationButtonLink
};

export interface IPaginationState {
  currentPage: number;
  lastPage: number;
  totalPages: number;
  pageLimit: number;
  totalRecords: number;
  pageNeighbours: number;
}

export interface IPaginationProps {
  totalRecords: number;
  pageLimit: number;
  pageNeighbours: number;
  hidePaginationLimiter?: boolean;
  linkType?: LinkType;
  onPageChanged: (paginationData: IPaginatorQuery) => void;
}

const Pagination = ({
  pageLimit = 1,
  pageNeighbours = 0,
  totalRecords = 0,
  hidePaginationLimiter = false,
  linkType = LinkType.BUTTON,
  onPageChanged,
  children
}: React.PropsWithChildren<IPaginationProps>) => {
  const [state, setState] = useState<IPaginationState>(() => ({
    currentPage: 1,
    lastPage: 1,
    totalPages: 0,
    pageLimit: pageLimit,
    totalRecords: totalRecords,
    pageNeighbours: pageNeighbours
  }));

  const gotoPage = useCallback((props) => {
    const currentPage = Math.max(1, Math.min(props.page, props.totalPages));
    onPageChanged({ page: currentPage, limit: props.pageLimit });
  }, [onPageChanged]);

  const handleClick = useCallback(currentPage => (evt) => {
    evt.preventDefault();
    setState(prevProps => ({ ...prevProps, currentPage, lastPage: prevProps.currentPage }));
  }, []);

  const handleMoveLeft = useCallback((evt) => {
    evt.preventDefault();
    setState(prevProps => ({ ...prevProps, lastPage: prevProps.currentPage, currentPage: state.currentPage - (state.pageNeighbours * 2) - 1 }));
  }, [state.currentPage, state.pageNeighbours]);

  const handleMoveRight = useCallback((evt) => {
    evt.preventDefault();
    setState(prevProps => ({ ...prevProps, lastPage: prevProps.currentPage, currentPage: state.currentPage + (state.pageNeighbours * 2) + 1 }));
  }, [state.currentPage, state.pageNeighbours]);

  const setPageLimit = useCallback((evt) => {
    evt.preventDefault();
    evt.persist();
    setState(prevProps => ({ ...prevProps, pageLimit: Number(evt.target.value), lastPage: prevProps.currentPage }));
  }, [setState]);

  const calculatedPages = useMemo(() => {
    if (totalRecords > 0) {
      return { neighbours: Math.max(0, Math.min(state.pageNeighbours, 2)), totalPages: Math.ceil(totalRecords / state.pageLimit), totalRecords };
    }
    return { neighbours: pageNeighbours, totalPages: 0, totalRecords };
  }, [totalRecords, state.pageLimit, pageNeighbours]); // eslint-disable-line

  const paginationPages = useMemo(() => {
    const { neighbours, totalPages } = calculatedPages;
    const totalNumbers = (neighbours * 2) + 3;
    const totalBlocks = totalNumbers + 2;
    if (totalPages > totalBlocks) {
      const startPage = Math.max(2, state.currentPage - neighbours);
      const endPage = Math.min(totalPages - 1, state.currentPage + neighbours);
      let pages = createPaginationRange({ from: startPage, to: endPage });
      const hasLeftSpill = startPage > 2;
      const hasRightSpill = (totalPages - endPage) > 1;
      const spillOffset = totalNumbers - (pages.length + 1);

      switch (true) {
        case (hasLeftSpill && !hasRightSpill): {
          const extraPages = createPaginationRange({ from: startPage - spillOffset, to: startPage - 1 });
          pages = [LEFT_PAGE, ...extraPages, ...pages];
          break;
        }

        case (!hasLeftSpill && hasRightSpill): {
          const extraPages = createPaginationRange({ from: endPage + 1, to: endPage + spillOffset });
          pages = [...pages, ...extraPages, RIGHT_PAGE];
          break;
        }

        case (hasLeftSpill && hasRightSpill):
        default: {
          pages = [LEFT_PAGE, ...pages, RIGHT_PAGE];
          break;
        }
      }

      return [1, ...pages, totalPages];
    }
    return createPaginationRange({ from: 1, to: totalPages });
  }, [calculatedPages, state.currentPage]);

  useEffect(() => {
    if ((totalRecords > 0 && totalRecords !== calculatedPages.totalRecords) || pageLimit !== state.pageLimit || state.currentPage !== state.lastPage) {
      gotoPage({ page: state.currentPage, pageLimit: state.pageLimit, ...calculatedPages });
    }
  }, [setState, totalRecords, calculatedPages, state]); // eslint-disable-line
  return (
    <>
      {children}
      <div className={css(styles.paginationNav)}>
        {!hidePaginationLimiter && (
          <select
            className={css(styles.pageLimiter)}
            value={state.pageLimit}
            onChange={setPageLimit}
          >
            <option value={1}>1</option>
            <option value={5}>5</option>
            <option value={10}>10</option>
            <option value={15}>15</option>
            <option value={20}>20</option>
            <option value={25}>25</option>
          </select>
        )}
        {(totalRecords > 0 && calculatedPages.totalPages > 1) && (
          <nav className={css(styles.navPagination)} aria-label="nav-pagination">
            <ul className={css(styles.pagination)}>
              {paginationPages.map((page, index) => {
                if (page === LEFT_PAGE) {
                  return (
                    <li key={index} className={css(styles.pageItem)}>
                      {typeMap[linkType]({ text: 'Prev', onClick: handleMoveLeft, page })}
                    </li>
                  );
                }

                if (page === RIGHT_PAGE) {
                  return (
                    <li key={index} className={css(styles.pageItem)}>
                      {typeMap[linkType]({ text: 'Prev', onClick: handleMoveRight, page })}
                    </li>
                  );
                }

                return (
                  <li key={index} className={state.currentPage === page ? `${css(styles.pageItemActive)}` : `${css(styles.pageItem)}`}>
                    {typeMap[linkType]({ page, text: page, onClick: handleClick(page), hideTick: true })}
                  </li>
                );
              })}
            </ul>
          </nav>
        )}
      </div>
    </>
  );
};

export default memo(Pagination);
