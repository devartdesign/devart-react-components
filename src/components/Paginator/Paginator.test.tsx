import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Paginator, { IPaginationProps } from './Paginator';

describe('Paginator Component', () => {
  global.console.error = () => null;
  let wrapper;
  const defaultProps: IPaginationProps = {
    page: 1,
    totalRecords: 40,
    pageLimit: 2,
    pageNeighbours: 2,
    hidePaginationLimiter: false,
    onPageChanged: jest.fn(),
  };

  describe('render', () => {
    beforeEach(() => {
      wrapper = shallow(<Paginator {...defaultProps} />);
    });

    it('should render with 20 items in page 1', () => {
      expect(renderer.create(wrapper)
        .toJSON())
        .toMatchSnapshot();
    });

    describe('componentDidUpdate', () => {
      beforeEach(() => {
        wrapper.setProps({
          totalRecords: 20
        });
      });

      it('should call onPageChanged callback from Props', () => {
        expect(defaultProps.onPageChanged).toHaveBeenCalledWith({
          currentPage: 1,
          totalPages: 10,
          pageLimit: 2,
          totalRecords: 20
        });
      });
    });

    describe('handleClick', () => {
      beforeEach(() => {
        const handleClick = wrapper.instance().handleClick(19);
        handleClick({ preventDefault: () => { /** */ } });
      });

      it('should navigate to page 19', () => {
        expect(renderer.create(wrapper)
          .toJSON())
          .toMatchSnapshot();
      });

      it('should call onPageChanged callback from', () => {
        expect(defaultProps.onPageChanged).toHaveBeenCalledWith({
          currentPage: 19,
          totalPages: 20,
          pageLimit: 2,
          totalRecords: 40
        });
      });
    });

    describe('handleMoveLeft', () => {
      beforeAll(() => {
        defaultProps.onPageChanged.mockReset();
        wrapper.instance().handleMoveLeft({ preventDefault: () => { /** */ } });
      });

      it('should call onPageChanged callback from Props', () => {
        expect(defaultProps.onPageChanged).toHaveBeenCalledWith({
          currentPage: 14,
          totalPages: 20,
          pageLimit: 2,
          totalRecords: 40
        });
      });
    });

    describe('handleMoveRight', () => {
      beforeAll(() => {
        defaultProps.onPageChanged.mockReset();
        wrapper.instance().handleMoveRight({ preventDefault: () => {} });
      });

      it('should call onPageChanged callback from Props', () => {
        expect(defaultProps.onPageChanged).toHaveBeenCalledWith({
          currentPage: 6,
          totalPages: 20,
          pageLimit: 2,
          totalRecords: 40
        });
      });
    });
  });
});
