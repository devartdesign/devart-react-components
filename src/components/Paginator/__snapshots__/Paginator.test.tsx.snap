// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator Component render handleClick should navigate to page 19 1`] = `
.paginationNav {
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  -webkit-box-align: center;
  -ms-flex-align: center;
  -ms-flex-preferred-size: 100%;
  -webkit-box-direction: normal;
  -webkit-box-orient: horizontal;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-basis: 100%;
  flex-basis: 100%;
  -webkit-align-items: center;
  align-items: center;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.pageLimiter {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  width: 60px;
  height: 30px;
  background-color: #fff;
  border-radius: 3px;
  border: 1px solid #CCC;
}

.navPagination {
  -ms-flex-preferred-size: 100%;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-flex-basis: 100%;
  flex-basis: 100%;
}

.pagination {
  -webkit-box-flex: 1;
  -ms-flex-positive: 1;
  background-color: #FFF;
  padding: 0 10px;
  margin: 10px 0;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-flex-grow: 1;
  flex-grow: 1;
  list-style-type: none;
}

.pageItem {
  -ms-flex-line-pack: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-content: center;
  align-content: center;
  width: 25px;
  height: 25px;
}

.pageItemButtonActive {
  cursor: pointer;
  border: none;
  background-color: #29434e;
  color: #FFF;
  width: 100%;
  height: 100%;
  padding: 0px;
}

.pageItemButton {
  cursor: pointer;
  border: none;
  background-color: #FFF;
  width: 100%;
  height: 100%;
  padding: 0px;
}

.pageItemButton:hover {
  background-color: #546e7a;
  color: #FFF;
}

<div
  className="paginationNav"
>
  <select
    className="pageLimiter"
    onChange={[Function]}
    value={2}
  >
    <option
      value={1}
    >
      1
    </option>
    <option
      value={5}
    >
      5
    </option>
    <option
      value={10}
    >
      10
    </option>
    <option
      value={15}
    >
      15
    </option>
    <option
      value={20}
    >
      20
    </option>
    <option
      value={25}
    >
      25
    </option>
  </select>
  <nav
    aria-label="nav-pagination"
    className="navPagination"
  >
    <ul
      className="pagination"
    >
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButtonActive"
          onClick={[Function]}
          type="button"
        >
          <span>
            1
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            2
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            3
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            4
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            5
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            6
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            7
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Â«
          </span>
          <span>
            Prev
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            20
          </span>
        </button>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`Paginator Component render should render with 20 items in page 1 1`] = `
.paginationNav {
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  -webkit-box-align: center;
  -ms-flex-align: center;
  -ms-flex-preferred-size: 100%;
  -webkit-box-direction: normal;
  -webkit-box-orient: horizontal;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-basis: 100%;
  flex-basis: 100%;
  -webkit-align-items: center;
  align-items: center;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.pageLimiter {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  width: 60px;
  height: 30px;
  background-color: #fff;
  border-radius: 3px;
  border: 1px solid #CCC;
}

.navPagination {
  -ms-flex-preferred-size: 100%;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-flex-basis: 100%;
  flex-basis: 100%;
}

.pagination {
  -webkit-box-flex: 1;
  -ms-flex-positive: 1;
  background-color: #FFF;
  padding: 0 10px;
  margin: 10px 0;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-flex-grow: 1;
  flex-grow: 1;
  list-style-type: none;
}

.pageItem {
  -ms-flex-line-pack: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-content: center;
  align-content: center;
  width: 25px;
  height: 25px;
}

.pageItemButtonActive {
  cursor: pointer;
  border: none;
  background-color: #29434e;
  color: #FFF;
  width: 100%;
  height: 100%;
  padding: 0px;
}

.pageItemButton {
  cursor: pointer;
  border: none;
  background-color: #FFF;
  width: 100%;
  height: 100%;
  padding: 0px;
}

.pageItemButton:hover {
  background-color: #546e7a;
  color: #FFF;
}

<div
  className="paginationNav"
>
  <select
    className="pageLimiter"
    onChange={[Function]}
    value={2}
  >
    <option
      value={1}
    >
      1
    </option>
    <option
      value={5}
    >
      5
    </option>
    <option
      value={10}
    >
      10
    </option>
    <option
      value={15}
    >
      15
    </option>
    <option
      value={20}
    >
      20
    </option>
    <option
      value={25}
    >
      25
    </option>
  </select>
  <nav
    aria-label="nav-pagination"
    className="navPagination"
  >
    <ul
      className="pagination"
    >
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButtonActive"
          onClick={[Function]}
          type="button"
        >
          <span>
            1
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            2
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            3
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            4
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            5
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            6
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            7
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Â«
          </span>
          <span>
            Prev
          </span>
        </button>
      </li>
      <li
        className="pageItem"
      >
        <button
          aria-label="Next"
          className="pageItemButton"
          onClick={[Function]}
          type="button"
        >
          <span>
            20
          </span>
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
